[IV-21-22] Objetivo 2

fjgallardo00 opened the pull request:
# Estructura de repositorio

* [x] ¿He seguido las mejores prácticas en nombre de las clases y ficheros y disposición de los mismos?

# Análisis de problema

* [x] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo que se ha creado?
* [x] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor, una entidad o un agregado?
* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo objeto valor que no dependa de ningún otro (y que sea la base de muchos otros)?

# Planificación y Programación

* [x] ¿Los issues representan un problema, y no una tarea?
* [x] ¿Todos los mensajes de commit explican el cambio, y no se limitan a repetir el nombre del fichero que se ha cambiado?
* [x] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están asignados a una HU?
* [x] ¿Todos los cambios en el código están asignados a un issue al que se referencia en un commit? Los issues sobre los que estoy trabajando, ¿han sido asignados al primer milestone?
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [x] ¿Son los milestones sobre los que estoy trabajando productos mínimamente viables? ¿O tengo que solicitar al product manager que cree milestones adicionales o precise de qué producto se trata?
* [x] ¿Se ha comprobado que el código entregado sea sintácticamente correcto?

fjgallardo00 committed:
Elección del lenguaje programación

fjgallardo00 committed:
Revert "Elección del lenguaje programación"

This reverts commit f165c206d05edf92da82aa4c7ebc470993cbb396.

fjgallardo00 committed:
Eleccion del lenguaje de programacion

fjgallardo00 committed:
Closes #11

Se elige Python como lenguaje de programación

fjgallardo00 committed:
Closes #11

Se arregla un pequeño fallo en el path de la entidad

mariavallejo20 commented:
Hola @fjgallardo00!! Se me ha asignado como revisora de tu objetivo 2, le he echado un vistazo a lo que tienes hasta ahora. Imagino que no es la entrega final (si me equivoco dímelo), avísame cuando lo tengas todo y te comento que me parece.
Como recomendación para lleva a acabo este objetivo te diría que leas:
* Los errores frecuentes del objetivo 2 de la [semana7](https://github.com/JJ/IV-22-23/blob/master/sesiones/semana-07.md) en adelante.
* Tener siempre en cuenta la cadena HU -> issue --> commit
* Revisar los objetivos de aprendizaje del objetivo 2 marcados en el [guión](http://jj.github.io/IV/documentos/proyecto/2.Entidad)

Ánimo y avísame cuando quieras que revise algo. 😇😁

fjgallardo00 commented:
Hola @mariavallejo20! Efectivamente, no es la entrega final. Hice el PR al repositorio de JJ para pasar todos los tests y continuar con el desarrollo. Gracias por los consejos, le echaré un vistazo de nuevo a todo. Cuando tenga todo volveré a avisar.

fjgallardo00 committed:
Avance en la tarea #13

Se ha definido la clase jugador y la clase equipo pero se necesitan más detalles para seguir con el desarrollo

fjgallardo00 commented:
@JJ en la lista de comprobación hay dos apartados que indican que hay que hacer documentación, pero en la entrega del objetivo no he visto nada de que tengamos que hacer documentación adicional sobre lo que hagamos (quizá no he leído bien, o lo he leído y lo pasé por alto). ¿Debemos de hacer una documentación explicando lo que hemos hecho o no sería necesario?

fjgallardo00 committed:
Avance en la tarea #13

Se ha definido la entidad quiniela y se han añadido más detalles en las entidades jugador y equipo

fjgallardo00 committed:
Avance en la tarea #13

Se han añadido más detalles en las entidades jugador y quiniela de acuerdo con lo especificado en la HU2 #8

JJ commented:
> @JJ en la lista de comprobación hay dos apartados que indican que hay que hacer documentación, pero en la entrega del objetivo no he visto nada de que tengamos que hacer documentación adicional sobre lo que hagamos (quizá no he leído bien, o lo he leído y lo pasé por alto). ¿Debemos de hacer una documentación explicando lo que hemos hecho o no sería necesario?

Eso, o no marcar la lista de comprobación.

fjgallardo00 commented:
@JJ @mariavallejo20 @LuisMart7 
Listo para revisión.

JJ commented:
Mira [este error frecuente en la semana 8](https://github.com/JJ/IV-22-23/blob/b23ba36a658b6ed0dc5851feaa5cd967cab8bd8b/sesiones/semana-08.md?plain=1#L115) sobre los mensajes de commit.

JJ commented:
Mira este error frecuente de la semana 8

https://github.com/JJ/IV-22-23/blob/b23ba36a658b6ed0dc5851feaa5cd967cab8bd8b/sesiones/semana-08.md?plain=1#L115

sobre los mensajes de *commit*

fjgallardo00 commented:
Poco más se le puede añadir a los mensajes de commit, salvo en el último de todos que sí podría haber puesto cuál era el pequeño fallo, pero por lo demás no sé que tendría que añadir. 

¿El resto de cosas está todo bien?

fjgallardo00 committed:
Closes #13

Se añade una pequeña explicación de la composición de las entidades y se arreglan pequeños fallos

JJ commented:
> Poco más se le puede añadir a los mensajes de commit, salvo en el último de todos que sí podría haber puesto cuál era el pequeño fallo, pero por lo demás no sé que tendría que añadir.
> 
> ¿El resto de cosas está todo bien?

No puedes añadir ni quitar nada a los commits. Lo que pones en la segunda línea debería estar en la primera, condensado.

En todo caso, es @LuisMart7 quien tiene que aprobarlo, antes de que lo mire yo.

LuisMart7 commented:
La implementación del modelo del problema la veo correcta. Las clases de equipo y jugador abarcan suficientes campos como para poner en marcha el siguiente milestone. La clase quiniela también es correcta pues a su vez contiene los enfrentamientos que se tienen en cuenta así como una estructura de datos donde plasmar la probabilidad de resultados. @JJ puede revisarlo que por mi parte lo veo correcto.

LuisMart7 reviewed (approved):
👍

fjgallardo00 commented:
@JJ Listo para revisión.

fjgallardo00 committed:
Closes #13. Crea objeto valor enfrentamiento y reestructuración de entidades

Se pasa la entidad jugador a ser un objeto valor y se crea la entidad estadisticas_jugador para ser la entidad

fjgallardo00 committed:
Closes #13. Se arregla el path de entidades de iv.yaml y se renombra la entidad estadisticas_jugador

fjgallardo00 commented:
@JJ Listo para revisión.

JJ commented:
Los dos últimos commits [f47245d](https://github.com/LuisMart7/ProBusiness/pull/12/commits/f47245d1d8deda3580b47785a4e97e229c110b14) y [c7e5686](https://github.com/LuisMart7/ProBusiness/pull/12/commits/c7e56862a82cbbba83a99d538e014961565f7e75) (al menos) no siguen las mejores prácticas.

JJ commented on iv.yaml:
```diff
@@ -0,0 +1,2 @@
+lenguaje: Python3
```
Python3 no es un lenguaje.

JJ commented on src/README.md:
```diff
@@ -0,0 +1,3 @@
+## Entidades
+
+Según lo hablado en [#13](https://github.com/LuisMart7/ProBusiness/issues/13), se ha decidido realizar tres entidades para realizar la tarea. Es necesario tener una entidad equipo con la que tenemos los datos generales de como va el equipo en la temporada, incluyendo también el presupuesto y los jugadores. Los jugadores también poseen una entidad para tener un seguimiento detallado sobre cada uno ...
```
"Lo hablado" no es como yo me referiría a un problema, lo que indica que claramente *no* habéis planteado un problema. Un issue tampoco es un diálogo: todo el diálogo tiene que ocurrir en el PR, en base a lo que se haya indicado en las historias de usuario.
Enunciar la decisión que habéis tomado tampoco explica el modelo del problema, y mucho menos es la forma de hacer *domain driven design*

JJ commented on src/enfrentamiento.py:
```diff
@@ -0,0 +1,15 @@
+        self.goles2 = goles2
+        self.fecha = fecha
+
+    def get_enfrentamiento(self):
```
No hay ninguna HU ni issue que pida un getter, y mucho menos en ese formato.

JJ commented on src/enfrentamiento.py:
```diff
@@ -0,0 +1,15 @@
+        self.equipo2 = equipo2
+        self.goles1 = goles1
+        self.goles2 = goles2
+        self.fecha = fecha
```
¿Hay alguna HU que pida la fecha? ¿Se usa en la lógica de negocio?
Hacer un modelo del problema, como hemos visto en el hackatón, es efectivamente ver qué es esencial para representar el problema y poder resolverlo. No se trata de tomar las palabras que aparecen en la descripción del problema y poner la idea previa que uno tenga del mismo.
En ese sentido, ¿hacen falta los goles? ¿O sólo la quiniela?

JJ commented on src/enfrentamiento.py:
```diff
@@ -0,0 +1,15 @@
+    def get_enfrentamiento(self):
+        return self.equipo1 + " - " + self.equipo2
+
+    def get_resultado(self):
```
Tampoco hay ninguna HU. Si la hubiera, seguramente no pediría el resultado sin los equipos enfrentados.

JJ commented on src/equipo.py:
```diff
@@ -0,0 +1,32 @@
+class Equipo:
+
+    def __init__(self, nombre):
```
Una de las cosas que pide el domain driven design es distinguir entre objetos valor y entidades para que las responsabilidades de gestionar cada cosa estén claras. Mirad [este error frecuente](https://github.com/JJ/IV-22-23/blob/master/sesiones/semana-12.md#objetivo-2) sobre este tema.
REcordad que, como vimos en el hackatón, hay que ir hasta encontrar el objeto-valor más simple con el que se pueda trabajar e ir construyendo a partir de ahí.

JJ commented on src/equipo.py:
```diff
@@ -0,0 +1,32 @@
+        self.partidos_jugados = 0
+        self.victorias = 0
+        self.empates = 0
+        self.derrotas = 0
```
No hay nada en la HU que diga que hace falta toda esta información. En todo caso, también es muy dudoso que esto se maneje desde un objeto que sólo contenga "equipo". Como he dicho antes, buscar palabras clave y poner la idea que tiene uno de esa palabra clave en una clase es un error. Hay que hacer un verdadero modelo del problema, atendiendo a cómo se va a usar desde la lógica de negocio.

JJ commented on src/equipo.py:
```diff
@@ -0,0 +1,32 @@
+        self.goles_a_favor = 0
+        self.goles_en_contra = 0
+        self.presupuesto_equipo = 0
+        self.posicion_liga = 1
```
Gestionar esto desde el propio objeto y no desde una entidad puede conducir a inconsistencias. El objeto, por ejemplo, no va a saber si hay otro que esté en la misma posición. Si no gestionas esta información desde una entidad, va a ser imposible que lo hagas correctamente.

JJ reviewed (changes requested):
Aparte de lo dicho, sobran todos los getters. Por favor, haz un verdadero análisis del problema partiendo de verdaderos problemas planteados en issues.

fjgallardo00 committed:
Closes #16. Se define a un jugador con nombre y posición

fjgallardo00 committed:
Closes #15. Se define a equipo con nombre y una lista de jugadores

fjgallardo00 committed:
Closes #14. Se define enfrentamiento entre dos equipos

fjgallardo00 commented:
@JJ @LuisMart7 Listo para revisión.

fjgallardo00 committed:
Closes #16. Se elimina jugador.py

fjgallardo00 committed:
Closes #15. Se elimina equipo.py

fjgallardo00 committed:
Closes #17. Se crea entidad jornada con enfrentamientos y resultados

fjgallardo00 committed:
Closes #18. Se crea entidad competicion con gestion de jornadas

fjgallardo00 committed:
Fix. Se cambia contenido iv.yaml

Se cambia el path de las entidades y se eliminan las que no sirven. Se actualiza el lenguaje

fjgallardo00 committed:
Refactor. Actualiza tipado de enfrentamiento

JJ commented on src/enfrentamiento.py:
```diff
@@ -0,0 +1,12 @@
+    equipo2 :str
+
+    # Se entiende por equipo1 como el equipo local y equipo2 como equipo visitante
+    def __init__(self, equipo1 :str, equipo2 :str):
```
¿El enfrentamiento no tiene el resultado?

fjgallardo00 replied:
Lo había puesto en la entidad superior, que es el array de resultados. Lo cambio y lo pongo dentro del objeto valor y lo quito de la entidad

JJ commented on src/jornada.py:
```diff
@@ -0,0 +1,14 @@
+
+    def __init__(self):
+        self.enfrentamiento = []
+        self.resultados = []
```
Os he dicho repetidamente que tratéis de almacenar las cosas para que se pueda trabajar con ellas de la forma más simple. ¿Te parece que almacenar en dos arrays en el mismo índice sea la mejor forma de hacerlo?

JJ reviewed (changes requested):
Desgraciadamente, creo que no se ha seguido una modelización correcta. Por favor, piensa de nuevo en la línea de lo que hemos hablado en clase.

fjgallardo00 commented:
@JJ Listo para revisión.

fjgallardo00 committed:
Fix gestión resultado

Enfrentamiento pasa a gestionar los goles de los partidos en lugar de jornada

JJ commented on iv.yaml:
```diff
@@ -0,0 +1,2 @@
+lenguaje: Python
+entidad: [ProBusiness/src/jornada.py, ProBusiness/src/competicion.py]
```
Los caminos a los ficheros no siguen las mejores prácticas. Por favor, consulta lo que se hace en caso de Python. En Python, los *paths* son nombres de módulos, y no quieres un módulo que se llame `src`

JJ commented on src/competicion.py:
```diff
@@ -0,0 +1,12 @@
+@dataclass
+class Competicion:
+
+    jornadas :list
```
Es una tontería, pero mira también las mejores prácticas en colocación de los `:`. En caso de duda, bájate `pylint` y aplícalo.

JJ commented on src/jornada.py:
```diff
@@ -0,0 +1,11 @@
+@dataclass
+class Jornada:
+
+    enfrentamiento :list # Debe de ser un pair que incluya la clase enfrentamiento y el resultado (1, X o 2)
```
¿Un solo enfrentamiento?

JJ reviewed (changes requested):
Como ya hablamos en general la semana pasada de las estructuras de datos, la objeción principal hoy son las mejores prácticas que debes seguir.

JJ commented:
Este commit: [0320737](https://github.com/LuisMart7/ProBusiness/pull/12/commits/0320737b631a99f18a03f39c322ffa6276a6177f) tampoco sigue las mejores prácticas, para empezar no se refiere a ningún issue.

JJ commented:
Adicionalmente, no veo ningún issue en el que se haya planteado el problema desde cero. Todos los issues que se habían creado aparentemente se han cerrado, así que no está claro que se siga la práctica sugerida para el desarrollo.

fjgallardo00 commented:
@JJ Listo para revisión

fjgallardo00 commented:
@JJ Listo para revisión.

JJ commented:
Tratar de no poned lo obvio en los mensajes de commit. Un mensaje de commit debe documentar las decisiones tomadas

JJ commented on competiciones/jornada.py:
```diff
@@ -0,0 +1,10 @@
+from dataclasses import dataclass
+
+@dataclass
```
¿Esta no es frozen?

JJ commented on competiciones/jornada.py:
```diff
@@ -0,0 +1,10 @@
+    enfrentamientos : list
+
+    def __init__(self):
+        self.enfrentamiento = []
```
Este no es el atributo declarado. Por favor, tened cierto cuidado cuando hacéis las cosas

JJ commented on iv.yaml:
```diff
@@ -0,0 +1,2 @@
+lenguaje: Python
+entidad: competiciones/jornada.py, competiciones/competicion.py
```
Esto sigue siendo YAML incorrecto.

JJ reviewed (changes requested):
Bastante cerca, salvo por errores de sintaxis.

fjgallardo00 committed:
Fix #19la sintaxis de las entidades y archivo yaml

Se cambia el path de las entidades en archivo yaml y se mejora la limpieza de codigo de entidades

fjgallardo00 committed:
Revert "Fix #19la sintaxis de las entidades y archivo yaml"

This revert ff02867. Se ha indicado mal la cabecera del commit

fjgallardo00 committed:
Fix #19 la sintaxis de las entidades y archivo yaml

Se cambia el path de las entidades en archivo yaml y se mejora la limpieza de codigo de entidades

fjgallardo00 committed:
Fix #19 en la entidad jornada

Se cambia el nombre de una variable para indicar la pluralidad de los datos que almacena

fjgallardo00 committed:
Closes #20. Se crea el modulo competiciones

fjgallardo00 committed:
Fix #20. Se indica el modulo competiciones en fichero yaml

fjgallardo00 committed:
Fix #19. Se cambia el path de entidades en fichero yaml

fjgallardo00 committed:
Fix #19. Se arregla sintaxis y tipado de entidades

Se arregla el tipado de las entidades para hacer buen uso de las dataclasses

fjgallardo00 commented:
@JJ Listo para revisión.

JJ commented on competiciones/competicion.py:
```diff
@@ -0,0 +1,13 @@
+from dataclasses import dataclass
```
Creo que todos estos caminos están mal. Los caminos a los componentes de un módulo tienen que ser relativos. Compruébalo en todo caso.

JJ commented on competiciones/competicion.py:
```diff
@@ -0,0 +1,13 @@
+class Competicion:
+
+    jornadas: List[Jornada]
+    clasificacion: dict
```
Como te he indicado en #18, `clasificacion` no queda claro qué problema resuelve. "Gestionar" no es la formulación de un problema. Tendrás que plantear el problema para que esto se pueda usar claramente desde la lógica de negocio. En concreto, no queda nada claro si esta clasificación es la de una jornada en concreto, y si lo es, de qué jornada. Tampoco se ha modelado ni se ha indicado por qué la clasificación está separada de la jornada, ni en qué ayuda eso a la lógica de negocio. Finalmente, tampoco se aclara por qué tiene que ser un diccionario y qué sería la clave y el valor de ese diccionario.

JJ commented on competiciones/jornada.py:
```diff
@@ -0,0 +1,12 @@
+    enfrentamientos: List[Enfrentamiento]
+
+    def __init__(self):
+        self.enfrentamientos = []
```
¿Qué se usa desde la lógica de negocio? ¿Los enfrentamientos? ¿Se usa el número de jornada? ¿Se usa la clasificación?

JJ reviewed (changes requested):
Creo que todavía tendrías que darle una vuelta a los problemas de los cuales tiene que emerger la modelización, y a la puesta en práctica de los mismos.

fjgallardo00 committed:
Fix #19. Se arregla sintaxis de variables list de entidades
